File: python/src/solana_pay_kit/protocols/x402/client/upto/payment.py
"""x402 ``upto`` client builder - ``payment-channel`` asset transfer method.

Parses an ``upto`` 402 challenge and builds the client authorization: a signed
channel ``open`` transaction (the deposit is the ceiling) plus the
``PAYMENT-SIGNATURE`` envelope. The client signs only its own (payer) slot in the
pull-style open; the advertised fee payer completes the fee-payer signature and
broadcasts. Mirrors the Go reference
(``go/protocols/x402/client/upto.go``).
"""

from __future__ import annotations

import base64
import json
import secrets
from collections.abc import Mapping
from typing import TYPE_CHECKING, Any, cast

from solders.hash import Hash  # type: ignore[import-untyped]
from solders.message import Message  # type: ignore[import-untyped]
from solders.pubkey import Pubkey  # type: ignore[import-untyped]

from solana_pay_kit._paycore.paymentchannels import (
    PAYMENT_CHANNELS_PROGRAM_ID,
    Distribution,
    OpenChannelParams,
    build_open_instruction,
    find_channel_pda,
)
from solana_pay_kit._paycore.solana import TOKEN_PROGRAM
from solana_pay_kit.protocols.x402.exact.verify import X402_VERSION
from solana_pay_kit.protocols.x402.upto.types import (
    UPTO_SCHEME,
    UptoPayload,
    UptoRequirements,
)

if TYPE_CHECKING:
    from solana_pay_kit.signer import LocalSigner

__all__ = [
    "parse_upto_challenge",
    "build_upto_payload",
    "encode_upto_header",
    "build_upto_header",
]

_U64_MAX = (1 << 64) - 1


def _parse_recent_slot(raw: Any) -> int:
    """Parse the server-provided ``extra.recentSlot`` (decimal string or number).

    The slot is emitted as a u64-as-string (matching the session challenge
    convention) but a plain JSON number is accepted too. Missing, negative, or
    out-of-range values raise ``ValueError``.
    """
    if raw is None or raw == "":
        raise ValueError("x402 client: requirement missing extra.recentSlot")
    if isinstance(raw, bool):
        raise ValueError(f"x402 client: invalid extra.recentSlot {raw!r}")
    if isinstance(raw, int):
        value = raw
    elif isinstance(raw, str) and raw.isascii() and raw.isdigit():
        value = int(raw, 10)
    else:
        raise ValueError(f"x402 client: invalid extra.recentSlot {raw!r}")
    if not 0 <= value <= _U64_MAX:
        raise ValueError(f"x402 client: extra.recentSlot {value} does not fit in u64")
    return value


def parse_upto_challenge(headers: Mapping[str, str], body: str | None = None) -> UptoRequirements | None:
    """Extract the ``upto`` requirement from a 402 challenge.

    Reads the base64 ``payment-required`` header, falling back to a JSON body,
    and returns the first ``accepts[]`` entry whose ``scheme`` is ``upto``.
    """
    envelope: Any = None
    raw = headers.get("payment-required") or headers.get("Payment-Required")
    if raw:
        try:
            envelope = json.loads(base64.b64decode(raw, validate=True))
        except Exception:  # noqa: BLE001
            return None
    elif body:
        try:
            envelope = json.loads(body)
        except Exception:  # noqa: BLE001
            return None
    else:
        return None
    if not isinstance(envelope, dict):
        return None
    accepts = cast("dict[str, Any]", envelope).get("accepts")
    if not isinstance(accepts, list):
        return None
    for entry in cast("list[Any]", accepts):
        if isinstance(entry, dict) and cast("dict[str, Any]", entry).get("scheme") == UPTO_SCHEME:
            return cast("UptoRequirements", entry)
    return None


def build_upto_payload(
    signer: LocalSigner,
    requirements: UptoRequirements,
    expires_at: int,
    nonce: str | None = None,
) -> UptoPayload:
    """Build the ``upto`` payload: a partially-signed channel ``open`` + metadata.

    The client (``signer``) is the channel payer. ``extra.feePayer`` is the
    transaction fee payer, rent payer, and zero-share channel payee (lifecycle
    authority); ``extra.receiverAuthorizer`` is the authorized voucher signer
    only. The open transaction is signed only in the client's payer slot
    (pull-style); the fee-payer slot is completed server-side.
    """
    extra = requirements["extra"]
    max_amount = int(requirements["amount"], 10)
    beneficiary = Pubkey.from_string(requirements["payTo"])
    mint = Pubkey.from_string(requirements["asset"])
    fee_payer_str = extra.get("feePayer")
    if not fee_payer_str:
        raise ValueError("x402 client: requirement missing extra.feePayer")
    fee_payer = Pubkey.from_string(fee_payer_str)
    receiver_authorizer_str = extra.get("receiverAuthorizer")
    if not receiver_authorizer_str:
        raise ValueError("x402 client: requirement missing extra.receiverAuthorizer")
    receiver_authorizer = Pubkey.from_string(receiver_authorizer_str)
    withdraw_delay = int(extra.get("withdrawDelay", 0))
    if withdraw_delay <= 0:
        raise ValueError("x402 client: requirement missing extra.withdrawDelay")
    # Always explicit: the payee seat is held by the facilitator (feePayer)
    # with a zero implicit remainder, so 100% of settled funds must be
    # assigned to payTo through the recipients list.
    recipients: list[Distribution] = [Distribution(recipient=beneficiary, bps=10_000)]
    program_id = Pubkey.from_string(PAYMENT_CHANNELS_PROGRAM_ID)
    token_program = Pubkey.from_string(extra.get("tokenProgram") or TOKEN_PROGRAM)

    blockhash_str = extra.get("recentBlockhash")
    if not blockhash_str:
        raise ValueError("x402 client: requirement missing extra.recentBlockhash")
    blockhash = Hash.from_string(blockhash_str)
    # The channel openSlot is server-provided as the challenge recentSlot
    # (like recentBlockhash); the client never fetches the slot itself. It is
    # a channel PDA seed and an openArgs field, so the requirement must carry it.
    open_slot = _parse_recent_slot(extra.get("recentSlot"))

    payer = Pubkey.from_string(signer.pubkey())
    salt = secrets.randbits(64)
    channel, _ = find_channel_pda(payer, fee_payer, mint, receiver_authorizer, salt, open_slot, program_id)
    open_ix = build_open_instruction(
        OpenChannelParams(
            payer=payer,
            rent_payer=fee_payer,
            payee=fee_payer,
            mint=mint,
            authorized_signer=receiver_authorizer,
            salt=salt,
            deposit=max_amount,
            grace_period=withdraw_delay,
            open_slot=open_slot,
            recipients=recipients,
            token_program=token_program,
            program_id=program_id,
        )
    )
    open_tx = _build_payer_signed_open(open_ix, fee_payer, payer, blockhash, signer)

    valid_after = int(extra.get("validAfter", 0))
    payload: UptoPayload = {
        "from": signer.pubkey(),
        "maxAmount": str(max_amount),
        "expiresAt": expires_at,
        "validAfter": valid_after,
        "nonce": str(salt),
        "channelId": str(channel),
        "deposit": str(max_amount),
        "authorizedSigner": receiver_authorizer_str,
        "openSlot": str(open_slot),
        "openTransaction": open_tx,
    }
    return payload


def encode_upto_header(requirements: UptoRequirements, payload: UptoPayload) -> str:
    """Encode the ``PAYMENT-SIGNATURE`` header (base64 JSON envelope)."""
    envelope = {
        "x402Version": X402_VERSION,
        "accepted": requirements,
        "payload": payload,
    }
    raw = json.dumps(envelope, separators=(",", ":")).encode("utf-8")
    return base64.b64encode(raw).decode("ascii")


def build_upto_header(
    signer: LocalSigner,
    requirements: UptoRequirements,
    expires_at: int,
    nonce: str | None = None,
) -> str:
    """Build the full ``PAYMENT-SIGNATURE`` header value for an ``upto`` retry."""
    payload = build_upto_payload(signer, requirements, expires_at, nonce)
    return encode_upto_header(requirements, payload)


def _build_payer_signed_open(
    open_ix: Any,
    fee_payer: Pubkey,
    payer: Pubkey,
    blockhash: Hash,
    signer: LocalSigner,
) -> str:
    """Assemble the open transaction with the advertised fee payer, signed only
    in the client's (payer) slot. Returns base64 wire."""
    message = Message.new_with_blockhash([open_ix], fee_payer, blockhash)
    account_keys = list(message.account_keys)
    num_required = int(message.header.num_required_signatures)
    try:
        payer_idx = account_keys.index(payer)
    except ValueError as exc:  # pragma: no cover - the open instruction always includes the payer
        raise ValueError("x402 client: payer not present in open transaction accounts") from exc
    message_bytes = bytes(message)
    payer_sig = bytes(signer.sign(message_bytes))
    signatures = bytearray(64 * num_required)
    signatures[payer_idx * 64 : payer_idx * 64 + 64] = payer_sig
    # Single-byte shortvec count (num_required is always small).
    wire = bytes([num_required]) + bytes(signatures) + message_bytes
    return base64.b64encode(wire).decode("ascii")

File: rust/crates/x402/src/client/upto/payment.rs
//! Client-side payment building for the x402 `upto` scheme (payment-channel).
//!
//! The client opens a channel whose `deposit` is the authorized maximum, with
//! `authorized_signer = operator` so the operator can settle the actual amount
//! with a single voucher. The client signs only the `open` transaction; the
//! operator broadcasts it and settles after metering.

use std::str::FromStr;

use solana_hash::Hash;
use solana_keychain::SolanaSigner;
use solana_pubkey::Pubkey;

use solana_pay_core::payment_channels as pc;

use crate::error::Error;
use crate::protocol::schemes::upto::{
    UptoPayload, UptoRequiredEnvelope, UptoRequirements, UptoSignatureEnvelope,
    PROFILE_PAYMENT_CHANNEL, UPTO_SCHEME,
};
use crate::{PAYMENT_REQUIRED_HEADER, X402_VERSION_V2};

/// Build an `upto` payload for a `payment-channel` requirement.
///
/// `expires_at` is the voucher/authorization deadline (Unix seconds); `nonce`
/// uniquely identifies this authorization. The requirement MUST carry
/// `extra.recentBlockhash` (the operator provides it in the 402 challenge).
pub async fn build_upto_payload(
    payer_signer: &dyn SolanaSigner,
    requirements: &UptoRequirements,
    expires_at: i64,
    nonce: impl Into<String>,
) -> Result<UptoPayload, Error> {
    if !requirements
        .extra
        .profiles
        .iter()
        .any(|p| p == PROFILE_PAYMENT_CHANNEL)
    {
        return Err(Error::Other(
            "requirement does not advertise the payment-channel profile".to_string(),
        ));
    }

    let max = requirements.max_amount()?;
    let payee = Pubkey::from_str(&requirements.pay_to)
        .map_err(|e| Error::Other(format!("invalid payTo: {e}")))?;
    let mint = Pubkey::from_str(&requirements.asset)
        .map_err(|e| Error::Other(format!("invalid asset mint: {e}")))?;
    let operator = Pubkey::from_str(&requirements.extra.fee_payer)
        .map_err(|e| Error::Other(format!("invalid feePayer: {e}")))?;
    let program_id = match &requirements.extra.channel_program {
        Some(value) => {
            Pubkey::from_str(value).map_err(|e| Error::Other(format!("invalid programId: {e}")))?
        }
        None => pc::default_program_id(),
    };
    let token_program = match &requirements.extra.token_program {
        Some(value) => Pubkey::from_str(value)
            .map_err(|e| Error::Other(format!("invalid tokenProgram: {e}")))?,
        None => Pubkey::from_str("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA")
            .expect("valid token program"),
    };
    let recent_blockhash = requirements
        .extra
        .recent_blockhash
        .as_deref()
        .ok_or_else(|| Error::Other("requirement missing extra.recentBlockhash".to_string()))?;
    let blockhash = Hash::from_str(recent_blockhash)
        .map_err(|e| Error::Other(format!("invalid recentBlockhash: {e}")))?;

    let salt = pc::random_salt();
    // operator is both the voucher signer (authorized_signer) and the fee payer.
    let open = pc::build_open_payment_channel_tx(
        payer_signer,
        &payee,
        &mint,
        &operator,
        salt,
        max,
        pc::DEFAULT_GRACE_PERIOD_SECONDS,
        vec![],
        &token_program,
        &program_id,
        &operator,
        blockhash,
    )
    .await?;

    Ok(UptoPayload {
        profile: PROFILE_PAYMENT_CHANNEL.to_string(),
        from: pc::pubkey_string(&payer_signer.pubkey()),
        max_amount: max.to_string(),
        expires_at,
        valid_after: requirements.extra.valid_after.unwrap_or(0),
        nonce: nonce.into(),
        channel_id: pc::pubkey_string(&open.channel_id),
        deposit: max.to_string(),
        authorized_signer: pc::pubkey_string(&operator),
        open_transaction: Some(open.transaction),
        signature: None,
    })
}

/// Wrap a payload in a `PAYMENT-SIGNATURE` envelope and base64-encode it.
pub fn encode_upto_header(
    requirements: &UptoRequirements,
    payload: UptoPayload,
) -> Result<String, Error> {
    // Emit the canonical x402 v2 shape: `{ x402Version, accepted, payload }`.
    // Per spec §5.2 the scheme/network live inside `accepted`, not at the
    // envelope level.
    let envelope = UptoSignatureEnvelope {
        x402_version: X402_VERSION_V2,
        accepted: serde_json::to_value(requirements)
            .map_err(|e| Error::Other(format!("upto accepted serialization failed: {e}")))?,
        payload,
    };
    let json = serde_json::to_string(&envelope)
        .map_err(|e| Error::Other(format!("upto envelope serialization failed: {e}")))?;
    Ok(base64::Engine::encode(
        &base64::engine::general_purpose::STANDARD,
        json.as_bytes(),
    ))
}

/// Build the full `PAYMENT-SIGNATURE` header value for an `upto` payment.
pub async fn build_upto_header(
    payer_signer: &dyn SolanaSigner,
    requirements: &UptoRequirements,
    expires_at: i64,
    nonce: impl Into<String>,
) -> Result<String, Error> {
    let payload = build_upto_payload(payer_signer, requirements, expires_at, nonce).await?;
    encode_upto_header(requirements, payload)
}

/// Parse a 402 `upto` challenge from a `PAYMENT-REQUIRED` header value or body.
pub fn parse_upto_challenge(
    headers: &[(String, String)],
    body: Option<&str>,
) -> Option<UptoRequirements> {
    let from_header = headers
        .iter()
        .find(|(name, _)| name.eq_ignore_ascii_case(PAYMENT_REQUIRED_HEADER))
        .and_then(|(_, value)| {
            base64::Engine::decode(&base64::engine::general_purpose::STANDARD, value).ok()
        })
        .and_then(|bytes| serde_json::from_slice::<UptoRequiredEnvelope>(&bytes).ok());

    let envelope = from_header
        .or_else(|| body.and_then(|b| serde_json::from_str::<UptoRequiredEnvelope>(b).ok()))?;

    envelope
        .accepts
        .into_iter()
        .find(|req| req.scheme == UPTO_SCHEME)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::protocol::schemes::upto::UptoExtra;

    const OPERATOR: &str = "9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin";

    fn requirements() -> UptoRequirements {
        UptoRequirements {
            scheme: UPTO_SCHEME.to_string(),
            network: "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1".to_string(),
            amount: "1000000".to_string(),
            asset: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v".to_string(),
            pay_to: "CXhrFZJLKqjzmP3sjYLcF4dTeXWKCy9e2SXXZ2Yo6MPY".to_string(),
            max_timeout_seconds: 300,
            extra: UptoExtra {
                profiles: vec![PROFILE_PAYMENT_CHANNEL.to_string()],
                decimals: Some(6),
                token_program: None,
                fee_payer: OPERATOR.to_string(),
                channel_program: None,
                recent_blockhash: Some(Hash::default().to_string()),
                last_valid_block_height: None,
                valid_after: None,
            },
        }
    }

    fn sample_payload() -> UptoPayload {
        UptoPayload {
            profile: PROFILE_PAYMENT_CHANNEL.to_string(),
            from: "CXhrFZJLKqjzmP3sjYLcF4dTeXWKCy9e2SXXZ2Yo6MPY".to_string(),
            max_amount: "1000000".to_string(),
            expires_at: 4_102_444_800,
            valid_after: 0,
            nonce: "n-1".to_string(),
            channel_id: "CXhrFZJLKqjzmP3sjYLcF4dTeXWKCy9e2SXXZ2Yo6MPY".to_string(),
            deposit: "1000000".to_string(),
            authorized_signer: OPERATOR.to_string(),
            open_transaction: Some("tx".to_string()),
            signature: None,
        }
    }

    #[test]
    fn encode_header_produces_upto_envelope() {
        let header = encode_upto_header(&requirements(), sample_payload()).unwrap();
        let bytes =
            base64::Engine::decode(&base64::engine::general_purpose::STANDARD, &header).unwrap();
        let envelope: UptoSignatureEnvelope = serde_json::from_slice(&bytes).unwrap();
        assert_eq!(
            envelope.accepted.get("scheme").and_then(|s| s.as_str()),
            Some(UPTO_SCHEME)
        );
        assert_eq!(envelope.payload.max_amount, "1000000");
        assert_eq!(envelope.x402_version, X402_VERSION_V2);
        assert!(envelope.accepted.is_object());
    }

    #[test]
    fn parse_challenge_reads_payment_required_header() {
        let envelope = UptoRequiredEnvelope {
            x402_version: X402_VERSION_V2,
            resource: None,
            accepts: vec![requirements()],
            error: None,
        };
        let json = serde_json::to_string(&envelope).unwrap();
        let value =
            base64::Engine::encode(&base64::engine::general_purpose::STANDARD, json.as_bytes());
        let headers = vec![(PAYMENT_REQUIRED_HEADER.to_string(), value)];

        let parsed = parse_upto_challenge(&headers, None).unwrap();
        assert_eq!(parsed.amount, "1000000");
        assert_eq!(parsed.extra.fee_payer, OPERATOR);
    }

    #[test]
    fn parse_challenge_returns_none_without_upto_offer() {
        assert!(parse_upto_challenge(&[], None).is_none());
    }
}
